ws-client/src/transport.rs

```diff
@@ -184,74 +195,132 @@ impl<'a> WsTransportClientBuilder<'a> {
-						}
+		let mut target = self.target;
+		let origin = self.origin_header.map(|o| [o]);
+		let mut err = None;
+
+		for _ in 0..self.max_redirections {
+			log::debug!("Connecting to target: {:?}", target);
+
+			for sockaddr in target.sockaddrs.clone() {
+				let tcp_stream = match connect(sockaddr, self.timeout, &target.host, &tls_connector).await {
+					Ok(stream) => stream,
+					Err(e) => {
+						log::debug!("Failed to connect to sockaddr: {:?}", sockaddr);
+						err = Some(Err(e));
+						continue;
 					}
+				};
+				let mut client = WsHandshakeClient::new(
+					BufReader::new(BufWriter::new(tcp_stream)),
+					&target.host_header,
+					&target.path_and_query,
+				);
+				if let Some(origin) = &origin {
+					client.set_headers(origin);
 				}
-				future::Either::Right((_, _)) => return Err(WsHandshakeError::Timeout(self.timeout)),
-			}
-		};
+				// Perform the initial handshake.
+				match client.handshake().await {
+					Ok(ServerResponse::Accepted { .. }) => {
+						log::info!("Connection established to target: {:?}", target);
+						let mut builder = client.into_builder();
+						builder.set_max_message_size(self.max_request_body_size as usize);
+						let (sender, receiver) = builder.finish();
+						return Ok((Sender { inner: sender }, Receiver { inner: receiver }));
+					}
 
-		log::debug!("Connecting to target: {:?}", self.target);
-		let mut client = WsRawClient::new(
-			BufReader::new(BufWriter::new(tcp_stream)),
-			&self.target.host_header,
-			&self.target.path_and_query,
-		);
-		if let Some(origin) = self.origin_header.as_ref() {
-			client.set_origin(origin);
+					Ok(ServerResponse::Rejected { status_code }) => {
+						log::debug!("Connection rejected: {:?}", status_code);
+						err = Some(Err(WsHandshakeError::Rejected { status_code }));
+					}
+					Ok(ServerResponse::Redirect { status_code, location }) => {
+						log::debug!("Redirection: status_code: {}, location: {}", status_code, location);
+						match location.parse::<Uri>() {
+							// redirection with absolute path => need to lookup.
+							Ok(uri) => {
+								// Absolute URI.
+								if uri.scheme().is_some() {
+									target = uri.try_into()?;
+									tls_connector = match target.mode {
+										Mode::Tls => {
+											let mut client_config = rustls::ClientConfig::default();
+											if let CertificateStore::Native = self.certificate_store {
+												client_config.root_store = rustls_native_certs::load_native_certs()
+													.map_err(|(_, e)| WsHandshakeError::CertificateStore(e))?;
+											}
+											Some(Arc::new(client_config).into())
+										}
+										Mode::Plain => None,
+									};
+									break;
+								}
+								// Relative URI.
+								else {
+									// Replace the entire path_and_query if `location` starts with `/` or `//`.
+									if location.starts_with('/') {
+										target.path_and_query = location;
+									} else {
+										match target.path_and_query.rfind('/') {
```

Comment:
//cc @maciejhirsz I had to do this instead of `Path::join` because on `Windows` the path is joined by `\` which is a faulty `URI relative resource`.

This will work for both `//` and `/` FWIW.